grammar.js

```diff
@@ -1,7 +1,201 @@
+        $.identifier,
+        $.constant,
+      ),
+
+    string: $ => choice($.quoted_string, $.percent_string),
+
+    chained_string: $ => seq($.string, repeat1($.string)),
+
+    quoted_string: $ => seq('"', repeat(/[^"]/), '"'),
+
+    // TODO: redo this lazy implementation
+    percent_string: $ =>
+      seq(
+        '%',
+        choice('(', '[', '{', '<', '|'),
+        repeat(/[^)\]}>|]/),
+        choice(')', ']', '}', '>', '|'),
+      ),
+
+    integer: $ =>
+      seq(
+        choice(
+          '0',
+          seq('0_', repeat(/[0-9_]/)),
+          seq('0b', repeat(choice('0', '1', '_'))),
+          seq('0o', repeat(/[0-7_]/)),
+          seq('0x', repeat(/[a-fA-F0-9_]/)),
+          seq(/[1-9]/, repeat(/[0-9_]/)),
+        ),
+        optional(
+          choice(
+            'i8',
+            'i16',
+            'i32',
+            'i64',
+            'i128',
+            'u8',
+            'u16',
+            'u32',
+            'u64',
+            'u128',
+          ),
+        ),
+      ),
+
+    float: $ =>
+      seq(
+        choice('0', '0_', /[1-9]/),
+        repeat(/[0-9_]/),
+        '.',
+        repeat1(/[0-9_]/),
+        optional(
+          seq(
+            choice('e', 'E'),
+            optional(choice('+', '-')),
+            /[0-9_]/,
+            repeat(/[0-9_]/),
+          ),
+        ),
+        optional(choice('f32', 'f64')),
+      ),
+
+    nil: $ => 'nil',
+
+    true: $ => 'true',
+    false: $ => 'false',
+
+    array: $ =>
+      seq(
+        '[',
+        optional(
+          choice(
+            $._statement,
+            seq($._statement, repeat(seq(',', $._statement))),
+          ),
+        ),
+        ']',
+        optional(seq('of', $.constant)),
+      ),
+
+    hash: $ =>
+      choice(
+        seq(
+          '{',
+          $._statement,
+          '=>',
+          $._statement,
+          repeat(seq(',', $._statement, '=>', $._statement)),
+          '}',
+          optional(
+            seq(
+              'of',
+              field('key_type', $.constant),
+              '=>',
+              field('value_type', $.constant),
+            ),
+          ),
+        ),
+        seq(
+          '{',
+          '}',
+          'of',
+          field('key_type', $.constant),
+          '=>',
+          field('value_type', $.constant),
+        ),
+      ),
+
+    proc: $ =>
+      seq(
+        '->',
+        optional(
+          seq(
+            '(',
+            field(
+              'params',
+              seq($.param, repeat(seq(',', $.param))),
+              optional(','),
+            ),
+            ')',
+          ),
+        ),
+        optional(field('return_type', seq(':', $.constant))),
+        $.block,
+      ),
+
+    param: $ =>
+      seq(
+        // repeat($.annotation),
+        optional(field('external_name', $.identifier)),
+        field('name', $.identifier), // support class/instance vars
+        optional(field('type', seq(':', $.constant))),
```

Comment:
Should the ':' and '=' be part of the type/default value, or should it be something like this? Don't know pros/cons either way so your solution may be perfectly fine, still learning tree-sitter myself.

```
optional(seq(':', field('type', $.constant)));
```

Reply:
Ah, good catch! Yeah, they don't need to be part of the field. There's a chance it could affect syntax highlighting so I'll change that.

Reply:
Fixed in 5257d0c0dec3b13dc041c53abdc9824c25429e87.